manager/scheduler/nodeinfo.go

```diff
@@ -94,3 +108,43 @@ func taskReservations(spec api.TaskSpec) (reservations api.Resources) {
 	}
 	return
 }
+
+// taskFailed records a task failure from a given service.
+func (nodeInfo *NodeInfo) taskFailed(ctx context.Context, serviceID string) {
+	if nodeInfo.recentFailures == nil {
```

Comment:
It was paranoia. We've had some issues in the scheduler before with nil pointer dereferences, due to things bypassing the constructor. It looks like this is not an issue anymore. Removed the check, and also the one in `countRecentFailures`.